README.md
# ali-data-proxy-lite 1.1.12 (支持rest, modulet, mtop, hsf, mysql)
## Install

```sh
$ npm install ali-data-proxy-lite
```
===

对restfull接口的更好支持:
*1.主要支持了application/json
*2.angular client支持（ng-proxy)

## 目录
 - [Why?]()
   - [工作原理图]() 
 - [使用前必读]()
 - [快速开始]()
   - [用例一 接口文件配置->引入接口配置文件->创建并使用proxy]()
   - [用例二 Proxy多接口配置及合并请求]()
   - [用例三 Proxy混合配置及依赖调用]()
   - [用例四 配置mock代理]()
   - [用例五 使用DataProxy拦截请求]()
   - [用例六 在浏览器端使用DataProxy]()
   - [用例七 代理带cookie的请求并且回写cookie]()
   - [用例八 使用DataProxy代理HSF的consumer功能]()
   - [用例九 使用DataProxy代理Mtop接口]()
   - [用例十 使用DataProxy操作Mysql数据库]()
   - [用例十一 使用DataProxy访问Modulet接口]()
 - [完整实例](demo/)
 - [配置文件详解]()
   - [interface.json 配置]()
   - [http 接口配置]()
   - [mtop 接口配置]()
   - [hsf 接口配置]()
   - [mysql 接口配置]()
 - [API]()
   - [DataProxy对象创建方式]()
   - [创建DataProxy对象时指定的profile相关形式]()
   - [DataProxy logger设置]()
   - [DataProxy对象方法]()
 - [如何使用DataProxy的Mock功能]()
   - [rule.json文件]()
   - [rule.json文件样式]()

## Why?
---
淘系的技术大背景下，必须依赖Java提供稳定的后端接口服务。在这样环境下，Node Server在实际应用中的一个主要作用即是代理(Proxy)功能。由于淘宝业务复杂，后端接口方式多种多样(MTop, Modulet, HSF...)。然而在使用Node开发web应用时，我们希望有一种统一方式访问这些代理资源的基础框架，为开发者屏蔽接口访问差异，同时提供友好简洁的数据接口使用方式。于是就有了 midway-dataproxy 这个构件。使用midway-dataproxy，可以提供如下好处：

1. 不同的开发者对于接口访问代码编写方式统一，含义清晰，降低维护难度。
2. 框架内部采用工厂+单例模式，实现接口一次配置多次复用。并且开发者可以随意定制组装自己的业务Proxy(依赖注入)。
3. 可以非常方便地实现线上，日常，预发环境的切换。
4. 内置[river-mock](http://gitlab.alibaba-inc.com/river/mock/tree/master)等mock引擎，提供mock数据非常方便。
5. 使用接口配置文件，对接口的依赖描述做统一的管理，避免散落在各个代码之中。
6. 支持浏览器端共享Proxy，浏览器端可以使用它做前端数据渲染。整个代理过程对浏览器透明。
7. 接口配置文件本身是结构化的描述文档，可以使用[river](http://gitlab.alibaba-inc.com/river/spec/tree/master)工具集合，自动生成文档。也可使用它做相关自动化接口测试，使整个开发过程形成一个闭环。

### DataProxy工作原理图及相关开发过程图览
---
![](http://img2.tbcdn.cn/L1/461/1/12bb633225499cdbba656335c3ec845dec7a92b4)

## 使用前必读
---
使用DataProxy之前，您需要在工程根目录下创建名为interface.json的配置文件。该文件定义了工程项目中所有需要使用到的接口集合(详细配置说明见后文)。定义之后，您可以在代码中按照需要引入不同的接口，创建与业务相关的Proxy对象。接口的定义和proxy其实是多对多的关系。也即一个接口可以被多个proxy使用，一个proxy可以使用多个接口。具体情况由创建proxy的方式来决定。下面用例中会从易到难指导您如何创建这些proxy。

## 快速开始
---

### 用例一 接口文件配置->引入接口配置文件->创建并使用proxy
* 第一步 配置接口文件命名为：interface_sample.json，并将其放在工程根目录下。
注意：整个项目有且只有一个接口配置文件，其interfaces字段下定义了多个接口。在本例中，仅仅配置了一个主搜接口。

```json
{
    "title": "xxxxx",
    "version": "1.0.0",
    "engine": "mockjs",
    "rulebase": "interfaceRules",
    "status": "prod",
    "enablePlugins": {
        "http": true
    },
    "interfaces": [ {
        "name": "主搜索接口",
        "id": "Search.getItems",
        "urls": {
            "prod": "http://s.m.taobao.com/client/search.do"
        }
    } ]
}
```

* 第二步 在代码中引入DataProxy模块，并且初始化引入接口配置文件（在实际项目中，引入初始化文件动作应伴随工程项目启动时完成，有且只有一次）

```js
// 引入模块
var DataProxy = require( 'dataproxy' ); 

// interface配置文件的绝对路径
var path = require('path').resolve( __dirname, './interface_sample.json' );

// 初始化引入接口配置文件  （注意：初始化工作有且只有一次）
DataProxy.init( path );
```

* 第三步 使用DataProxy

```javascript
// 创建proxy
var searchProxy = new DataProxy( {
    searchItems: 'Search.getItems'  // 自定义方法名: 配置文件中的定义的接口ID
} );
// 或者这样创建: var searchProxy = new DataProxy( 'Search.getItems' ); 此时getItems 会作为方法名

// 使用proxy, 注意: 调用方法所需要的参数即为实际接口所需要的参数。
searchProxy.searchItems( { keyword: 'iphone6' } )
    // !注意 必须调用 done 方法指定回调函数，来取得上面异步调用searchItems获得的数据!
    .done( function( data ) {
        console.log( data );
    } )
    .error( function( err ) {
        console.log( err );
    } );
```

### 用例二 Proxy多接口配置及合并请求
* 配置

```json
{   // 头部配置省略...
    "enablePlugins": {
        "http": true
    },
    "interfaces": [ {
        "name": "主搜索搜索接口",
        "id": "Search.list",
        "urls": {
            "prod": "http://s.m.taobao.com/search.do"
        }
    }, {
        "name": "热词推荐接口",
        "id": "Search.suggest",
        "urls": {
            "prod": "http://suggest.taobao.com/sug"
        }
    }, {
        "name": "导航获取接口",
        "id": "Search.getNav",
        "urls": {
            "prod": "http://s.m.taobao.com/client/search.do"
        }
    } ]
}
```

* 代码

```js
// 更多创建方式，请参考后文API
var proxy = new DataProxy( 'Search.*' );

// 调用自动生成的不同方法
proxy.list( { keyword: 'iphone6' } )
    .done( function( data ) {
        console.log( data );
    } );

proxy.suggest( { q: '女' } )
    .done( function( data ) {
        console.log( data );
    } )
    .error( function( err ) {
        console.log( err );
    } );

// 合并请求
proxy.suggest( { q: '女' } )
    .list( { keyword: 'iphone6' } )
    .getNav( { key: '流行服装' } )
    .done( function( data1, data2, data3 ) {
        // 参数顺序与方法调用顺序一致
        console.log( data1, data2, data3 );
    } );
```

### 用例三 Proxy混合配置及依赖调用

* 配置

```json
{   // 头部配置省略...
    "enablePlugins": {
        "http": true
    },
    "interfaces": [ {
        "name": "用户信息查询接口",
        "id": "Session.getUser",
        "type": "http",
        "urls": {
            "prod": "http://taobao.com/getUser.do"
        }
    }, {
        "name": "订单获取接口",
        "id": "Order.getOrder",
        "type": "http",
        "urls": {
            "prod": "http://taobao.com/getOrder"
        }
    } ]
}
```

* 代码

``` js
var proxy = new DataProxy( {
    getUser: 'Session.getUser',
    getMyOrderList: 'Order.getOrder'
} );
// 先获得用户id，然后再根据id号获得订单列表
proxy.getUser( { sid: 'fdkaldjfgsakls0322yf8' } )
    .done( function( data ) {
        var uid = data.uid;
        this.getMyOrderList( { id: uid } )
            .done( function( data ) {
                console.log( data );
            } )
            .fail( function( err ) {
                console.error( err );
            } );
    } )
    .fail( function( err ) {
        console.error( err );
    } );

```

### 用例四 配置mock代理
* 第一步 在相关接口配置段落中启用mock

```json
{
    "title": "pad淘宝数据接口定义",
    "version": "1.0.0",
    "engine": "river-mock",                       <-- 指定mock引擎
    "rulebase": "interfaceRules",   <-- 指定存放相关mock规则文件的目录名，约定位置与interface.json文件存放在同一目录，默认为 interfaceRules
    "status": "prod",
    "enablePlugins": {
        "http": true
    },
    "interfaces": [ {
        "name": "主搜索接口",
        "id": "Search.getItems",
        "ruleFile": "Search.getItems.rule.json",  <-- 指定数据mock规则文件名，如果不配置，则将默认设置为 id + '.rule.json'
        "urls": {
            "prod": "http://s.m.taobao.com/client/search.do",
            "prep": "http://s.m.taobao.com/client/search.do",
            "daily": "http://daily.taobao.net/client/search.do"
        },
        "status": "mock"                            <-- 启用mock状态，覆盖全局status
    } ]
}
```

* 第二步 添加接口对应的规则文件到ruleBase(interfaceRules)指定的文件夹。mock数据规则请参考[river-mock](http://gitlab.alibaba-inc.com/river/mock/tree/master)和[mockjs](http://mockjs.com)。启动程序后，DataProxy即返回相关mock数据。

### 用例五 使用DataProxy拦截请求

```js
var app = require( 'connect' )();
var DataProxy = require( 'dataproxy' );
var path = require('path').resolve( __dirname, './interface_sample.json' );
DataProxy.init( path );

// 指定需要拦截的路径
app.use( '/proxy', DataProxy.Interceptor );

// 此时可直接通过浏览器访问 /proxy/[interfaceid] 调用相关接口(如果该接口定义中配置了 intercepted = false, 则无法访问)
```

### 用例六 在浏览器端使用DataProxy
* 第一步 按照用例二配置接口文件

* 第二步 按照用例五 启用拦截功能

* 第三步 在浏览器端使用DataProxy

```html
<!-- 引入dataproxy模块，为方便起见直接引入，而非采用KISSY标准包配置引入。但该模块本身是由KISSY封装的标准模块-->
<script src="dataproxy-client.js" ></script>
```

```html
<script type="text/javascript">
    KISSY.use( "dataproxy", function( S, DataProxy ) {
        // !配置基础路径，该路径与第二步中配置的拦截路径一致!
        // 且全局配置有且只有一次！
        DataProxy.configBase( '/proxy/' );

        // 创建proxy
        var searchProxy = DataProxy.create( 'Search.*' );
        searchProxy
            .list( { q: 'ihpone6' } )
            .list( { q: '冲锋衣' } )
            .suggest( { q: 'i' } )
            .getNav( { q: '滑板' } )
            .done( function( data1, data2, data3, data4 ) {
                console.log( {
                    "list_ihpone6": data1,
                    "list_冲锋衣": data2,
                    "suggest_i": data3,
                    "getNav_滑板": data4
                } );
            } )
            .fail( function( err ) {
                console.error( err );
            } );

    } );
</script>
```

### 用例七 代理带cookie的请求并且回写cookie (注：请求是否需要带cookie或者回写取决于接口提供者)

* 关键代码(app 由express创建)

```js
app.get( '/getMycart', function( req, res ) {
    var cookie = req.headers.cookie;
    var cart = DataProxy.create( 'Cart.*' );
    cart.getMyCart()
        // 在调用done之前带上cookie
        .withCookie( cookie )
        // done 回调函数中最后一个参数总是回写的cookie，不需要回写时可以忽略
        .done( function( data , cookies ) {
           res.send( data ); 
        } )
        // 回写cookie
        .rewriteCookie( res ) // <-参数也可以是自定义回写函数，该回写函数的参数即待回写的cookie。见API .rewriteCookie 方法说明。
        .error( function( err ) {
            res.send( 500, err );
        } );
} );
```

### 用例八 使用DataProxy代理HSF的consumer功能

* 第一步，在interface.json文件中配置hsf interface

``` js
{
    "title": "xxxx",
    "version": "1.0.0",                      
    "engine": "river-mock",                  
    "status": "prod",
    "enablePlugins": {
        "hsf": true
    },                        
    "hsf": {                              
        "configServers": {
            "prod": "commonconfig.config-host.taobao.com",
            "daily": "10.232.16.8",          
            "prep": "172.23.226.84"
        }
    },
    "interfaces": [ {
        "type": "hsf",
        "id": "UserService.append",
        "service": "com.taobao.uic.common.service.userdata.UicDataService",
        "methodName": "insertData"
    } ]
}
```

* 第二步，使用DataProxy

```js
var UserService = DataProxy.create( 'UserService.*' );

var user = [];
user.push( require( 'js-to-java' ).Long(456) );
user.push('data-info-userdata');
user.push('gongyangyu');
user.push('langneng');

UserService.append( user )
    .done( function( result ) {
         console.log( result );
    } )
    .error( function( err ) {
         console.log( err );
    } );
```

* `补充说明`：
 - 1. 使用DataProxy调用hsf服务时，其所需要的参数类型及结构由hsf服务提供者决定。当服务提供者使用java发布service时，其参数要求有确定的数据类型，关于java和js之间的数据类型转换，请参考[node-hsf之Java 对象与 Node 的对应关系以及调用方法章节](http://gitlab.alibaba-inc.com/node/node-hsf/tree/master)。推荐使用[js-to-java](https://github.com/node-modules/js-to-java)来辅助编写Java对象。关于hsf的相关说明请参考[HSF项目说明](http://confluence.taobao.ali.com/pages/viewpage.action?pageId=819280)。此外，您可以使用[HSF服务治理](http://ops.jm.taobao.net/service-manager/service_search/index.htm?envType=daily)查询相关服务接口，并且使用[Nexus](http://mvnrepo.taobao.ali.com/nexus/index.html#welcome)查询并获得相关jar包。
 - 2. hsf 的mock方法与其他类型代理无区别，且可以参照例六直接在浏览器端调用hsf service

### 用例九 使用DataProxy代理Mtop接口
* 第一步，在interface.json文件中配置mtop interface

``` js
{
    "title": "pad淘宝项目数据接口集合定义",
    "version": "1.0.0",                      
    "engine": "river-mock",                  
    "status": "prod",
    "enablePlugins": {
        "http": true,
        "mtop": true
    },
    "interfaces":[ {
        "id": "Detail.getTaobaoDyn",
        "version": "1.0",
        "type": "mtop",
        "api": "com.taobao.detail.getTaobaoDyn"
    } ]
}
```

* 第二步，使用DataProxy

```js
var detail = DataProxy.create( 'Detail.*' );
detail.getTaobaoDyn( {'itemNumId': 37194529489} )
    // 在Node端访问Mtop必须带上cookie，cookie一般取自req.headers.cookie 字段
    .withCookie( req.headers.cookie )
    .done( function( data ) {
        console.log( data );
    } ).fail( function( err ) {
        console.log( err );
    } );
```

* `补充说明`：mtop 的mock方法及在浏览器端使用方法与其他无差别。

### 用例十 使用DataProxy操作mysql数据库
* 配置

```js
{
    "title": "xxxx",
    "version": "1.0.0",                      
    "engine": "river-mock",                  
    "status": "prod",
    "enablePlugins": {
        "mysql": true
    },
    "mysql": {
        "pools": {
            "test": {
                "host": "127.0.0.1",
                "port": 3306,
                "user": "user1",
                "database": "testdb",
                "password": "passw0rd",
                "connectionLimit": 100
            }
        },
        "defaultPool": "test"
    },
    "interfaces":[ {
        "id": "UserService.getUser",
        "type": "mysql",
        "sql": "SELECT * FROM USERINFO WHERE ID = ?"
    }, {
        "id": "UserService.addUser",
        "type": "mysql",
        "sql": "INSERT INTO USERINFO VALUES(?, ?, ?, ?)"
    }, {
        "id": "UserService.updateUser",
        "type": "mysql",
        "sql": "UPDATE USERINFO SET ?? = ? WHERE ID = ?"
    } ]
}
```

* 代码

```js
var userService = DataProxy.create( 'UserService.*' );

userService
    .getUser( ['user1'] )
    .addUser( ['user2', '男', '1999-01-01', 'passw0rd'] )
    .updateUser( ['GENDER', '女', 'user3' ] )
    .done( function( result1, result2, result3 ) {
        console.log.( result1, result2, result3 );
    } )
    .fail( function( err ) {
        console.error( err );
    } );
```

### 用例十一 使用DataProxy访问Modulet接口
* 第一步，在interface.json文件中配置modulet interface

``` js
{
    "title": "我的淘宝改版接口定义",
    "version": "1.0.0",                      
    "engine": "river-mock",                  
    "status": "prod",
    "enablePlugins": {
        "http": true,
        "modulet": true
    },
    "modulet": {
        "urls": {
            prod: "http://i.daily.taobao.net/module.do"
        }
    },
    "interfaces":[ {
        "id": "ITaobao.getMyFavourite",
        "version": "1.0",
        "type": "modulet",
        "dataOnly": true,                    // <-- 接口请求成功时，只返回不包含协议信息的数据。
        "moduletName": "favourite"           // <-- 请求的模块名
    } ]
}
```

* 第二步，使用DataProxy

```js
var iTaobao = DataProxy.create( 'ITaobao.*' );
iTaobao.getMyFavourite( {uid: 'xxxx'} )
    .withCookie( req.headers.cookie )
    .done( function( data ) {
        console.log( data );
    } )
    .fail( function( err ) {
        console.log( err );
    } );
```

### 完整实例请查看 [demo](demo/)

## 配置文件详解
---
### interface.json 配置结构

``` js
{
    "title": "xxx数据接口定义",                // [必填][string] 接口文档标题
    "version": "1.0.0",                      // [必填][string] 版本号
    "engine": "river-mock",                  // [选填][string] mock引擎，取值可以是river-mock和mockjs。不需要mock数据时可以不配置
    "rulebase": "interfaceRules",            // [选填][string] mock规则文件夹名称。不需要mock数据时可以不配置。约定该文件夹与
                                             // interface.json配置文件位于同一文件夹。默认为interfaceRules
    "status": "prod",                        // [必填][string] 全局代理状态，取值只能是 interface.urls中出现过的键值或者mock
    "enablePlugins": {                       // [选填][object] 插件启用配置，配置内容与[lib/plugins]所集成的插件一致。注意插件之间的依赖关系
        "http": true,                        // [选填][boolean] 是否启用http代理插件。设置为false时则不会做相应地插件初始化工作。默认false
                                             // 注意：如果禁用http代理会影响其他继承该插件实现的代理，比如mtop
        "mtop": true,                        // 同上
        "modulet": true,                     // 同上
        "mysql": true,                       // 同上
        "hsf": false                         // 同上
    },
    /* http 接口代理配置 */
    "http": { // httpClient 相关配置，不需要是可以不配置，`注意`：该配置项会影响所有其他基于http接口实现的子接口配置效果，比如mtop。
        "maxkSockets": 1000,                 // [选填][number] 最大socket链接数，默认1000。建议与后端http server所支持的并发链接数一致。
        "keepAliveMsecs": 3000               // [选填][number] 发送TCP keepAlive包的间隔时长。默认为3000。注意需要考虑Node应用实际部署
                                             // 的情况。建议当Node与后端服务（可以是Java）部署在同一台机器上时，设置为3000。分开部署时如果网络

CHANGELOG.md
## v1.1.6
* 修复1.1.5 改动引起的bug！

## v1.1.5
* 当http请求返回状态码为201或者202时，作为正确响应处理而非报错。
## v1.1.4
* http支持PUT方式

## v1.1.3
* 当用户配置的url被解析之后，path为空时会及时抛出异常，告知用户url不可用。

## v1.1.2
* 增加自定义回调错误标识，当fail或error函数被调用时，传入的err如果statusCode = -1，则表示为用户自己写的回调函数中存在错误，而非dataproxy请求数据时发生的错误。

## v1.1.1
* 去掉console

## v1.1.0
* 支持modulet
* 支持模拟数据时，优先启用以mock.json结尾的river-spec文件

## v1.0.10
* 修复了某些场景超时情况下request对象未创建而导致的undefined错误。

## v1.0.9
* 删除了一行多余的参数设置代码。

## v1.0.8
* 支持运行时动态添加interface options来创建DataProxy。

## v1.0.7
* 修复mysql plugin 重复报错的bug。

## v1.0.6
* 修复mysql plugin 不能正确报错的bug
* 提供在初始化DataProxy时，手工开启mock或者mockerr的快捷方式，方便开发调试

## v1.0.5
* 提供 rewriteCookie 方法，使之与 withCookie 方法调用对称。

## v1.0.4
* Bugfix 直接传入profile对象时也能正确的更新_config字段

## v1.0.3
* DataProxy 初始化参数bugfix

## v1.0.2
* 去掉对配置文件支持引入变量的解析功能
* 接入Midway-Framework插件标准

## v1.0.1
* 修改enablePlugins配置逻辑，默认为false，需要用户在配置文件中显式地启用plugin

## v1.0.0
* ModelProxy 正式更名为DataProxy 并且发布1.0版

## v0.6.2
* http plugin 支持url的path中带/:\w+/变量的解析，如url = 'projects/:id/files/:name'， 其中:id 和 :name在发送请求之前会被参数params.id和params.name替换。

## v0.6.1
* 增加捕捉导致 URIError 的异常字符的 log，直接输出到控制台

## v0.6.0
* 在http.js中querystringify方法内加try/catch 防止URIError和JSON循环引用问题

## v0.6.0-beta
* 修复了httpproxy重构之后关于res变量引用位置错误的bug
* 支持ModelProxy在node端调用mysql数据库
* 支持ModelProxy在浏览器端调用mysql接口
* 完善了mysql proxy测试用例

## v0.5.1-beta
* interface新增 enablePlugins 配置，方便用户通过配置启用或者禁用各种接口代理插件。
* 重构proxy.js，改变各种类型的接口拦截方式，如果为mock状态，则统一由proxy class代理拦截。

## v0.5.0
* http request final callback 之前增加判断是否结束，以防止重复callback

## v0.5.0-alpha1
* 修复了浏览器端调用mtop接口失败的bug

## v0.5.0-alpha
* 支持ModelProxy在node端访问mtop接口
* 支持ModelProxy在浏览器端访问mtop接口

## v0.4.1
* 暂时移除对hsf-protocol-cpp的依赖，以适应aone开发机部署环境。

## v0.4.0
* 收藏夹试点项目压测通过，版本已稳定。

## v0.4.0-alpha6
* 修改hsf timeout参数单位以及group默认值

## v0.4.0-alpha5
* 支持keepAliveMsecs参数配置，以适应不同部署场景下由于该参数设置不当而发生频繁的socket ECONNRESET错误

## v0.4.0-alpha4
* http请求失败的错误提示中增加rid，方便定位。

## v0.4.0-alpha3
* 增加http maxSockets配置文件
* 增加 http proxy callback 标记，避免重复callback。
* 修改错误提示中interface id显示成url的错误。
* http proxy 每次请求增加 rid 以标识唯一性

## v0.4.0-alpha2
* 修复hsf插件加载的bug
* 修复cookie专递失效的bug

## v0.4.0-alpha1
* 修复http timeout引起的socket hang up的bug

## v0.4.0-alpha
* hsfproxy功能实现
* hsfproxy拦截器实现
* hsfproxy浏览器端调用实现

## v0.3.2-beta
* 修复request interceptor getHeader的bug
* 重构InterfaceManager，采用Class方式对外提供方法。

## v0.3.2-alpha
* 增强statusCode不为200的报错内容

## v0.3.1
* 增强处理StatusCode 不为200的情况
* 增强错误提示，每次请求代理失败，需要明确对应的interface id以及请求的url，以方便错误定位

## v0.3.0
* 修复querystring 与version参数拼接的bug。
* 修复done内方法异常捕获error变量未定义的bug。

## v0.3.0-beta1
* 修复interface Rule路径读取bug。

## v0.3.0-beta
* Proxy 插件化实现。

## v0.3.0-alpha-2
* 升级river-mock。

## v0.3.0-alpha
* proxy底层采用继承方式重构。
* 支持捕获customized code异常。

## v0.2.8
* 支持interface配置文件变量引用。
* 重载ModelProxy.init( path, variables )。variables参数为开发者传入的用于解析interface配置文件中出现的变量的变量对象。


lib/plugins/http.js
/**
 * Http Proxy Class
 */

var util = require('util')
    , fs = require('fs')
    , url = require('url')
    , qs = require('querystring')
    , http = require('http')
    , iconv = require('iconv-lite')
    , Agent = require('agentkeepalive')
    , Constant = require('../constant');

// Agent will be initialized when HttpProxy.init is call
var keepAliveAgent;

var ProxyBase = require('../proxybase');

var httpConfig = {};

// HttpProxy Constructor
function HttpProxy(options) {
    this._opt = options || {};

    var urls = this._opt.urls || {};

    if (this._opt.status === Constant.STATUS_MOCK
        || this._opt.status === Constant.STATUS_MOCK_ERR) {
        return;
    }
    var currUrl = urls[this._opt.status];

    if (!currUrl) {
        throw new Error('No url can be proxied! InterfaceId = ' + options.id);
    }
    this._opt.currUrl = currUrl;
    var urlObj = url.parse(currUrl);
    this._opt.hostname = urlObj.hostname;
    this._opt.port = urlObj.port || 80;
    if (typeof urlObj.path !== 'string') {
        throw new Error('Url is unavailable! InterfaceId = ' + options.id);
    }
    this._opt.path = urlObj.path
    + ( urlObj.path.indexOf('?') !== -1 ? '&' : '?' )
    + '_version=' + this._opt.version + '&';
}

// Inherits ProxyBase
util.inherits(HttpProxy, ProxyBase);

// @override requestReal function of ProxyBase which is defined in proxybase.js
HttpProxy.prototype.requestReal = function (params, callback, errCallback, cookie) {
    if (this._opt.isCookieNeeded === true && cookie === undefined) {
        errCallback(new Error('This request need cookie, you must set a cookie for'
        + ' it before request. interfaceId = ' + this._opt.id));
        return;
    }

    var self = this, isCompleted = false;
    var rid = new Date().getTime() + '' + Math.round(Math.random() * 1000);

    var options = {
        hostname: self._opt.hostname,
        port: self._opt.port,
        path: self._parsePath(self._opt.path, params) + '_rid=' + rid + '&',
        method: self._opt.method,
        headers: {'Cookie': cookie},
        agent: keepAliveAgent,
        keepAlive: true
    };

    var querystring;

    if (self._opt.method === Constant.POST || self._opt.method === Constant.PUT) {
        if (self._opt.contentType && self._opt.contentType === 'application/json') {
            options.headers['Content-Type'] = Constant.MediaType.APPLICATION_JSON;
            querystring = JSON.stringify(params);
            console.log('path:' + options.path)
        } else {
            options.headers['Content-Type'] = 'application/x-www-form-urlencoded';
            querystring = self._queryStringify(params)
        }
        options.headers['Content-Length'] = querystring.length;

    } else if (self._opt.method === Constant.GET) {
        querystring = self._queryStringify(params);
        options.path += querystring;
    }
    var timer = setTimeout(function () {
        if (req) {
            req.abort();
        }
        var e = new Error('Timeout. rid = ' + rid
        + ', interfaceId = ' + self._opt.id
        + ', url = ' + self._opt.currUrl);
        errCallback(e);
        isCompleted = true;
    }, self._opt.timeout);

    var req = http.request(options, function (res) {
        if (isCompleted) return;
        var source = [], size = 0;
        res.on('data', function (chunk) {
            source.push(chunk);
            size += chunk.length;
        });
        res.on('end', function () {
            clearTimeout(timer);
            var buffer = Buffer.concat(source, size);
            if (res.statusCode !== 200 && res.statusCode !== 201 && res.statusCode !== 202) {
                var e = new Error('Response error. rid = ' + rid
                + ', interfaceId = ' + self._opt.id
                + ', url = ' + self._opt.currUrl
                + ', statusCode = ' + res.statusCode);
                e.statusCode = res.statusCode;
                e.responseText = buffer.toString();
                ( isCompleted = true ) && errCallback(e);
                return;
            }
            try {
                var result = self._opt.encoding === Constant.ENCODING_RAW
                    ? buffer
                    : ( self._opt.dataType !== Constant.JSON && self._opt.dataType !== Constant.JSONP
                    ? iconv.fromEncoding(buffer, self._opt.encoding)
                    : JSON.parse(iconv.fromEncoding(buffer, self._opt.encoding)) );
            } catch (e) {
                errCallback(new Error('The result has syntax error. interfaceId = '
                + self._opt.id + ', url = ' + self._opt.currUrl + '. Caused By: ' + e.message));
                return;
            }

            isCompleted || callback(result, res.headers['set-cookie']);
            isCompleted = true;
        });

        res.on('error', function (e) {
            if (isCompleted) return;
            clearTimeout(timer);
            e.message = 'Response failed. rid = ' + rid + ', interfaceId = ' + self._opt.id + ', url = '
            + self._opt.currUrl + '. Caused By: ' + e.message + ' code = ' + e.code;
            errCallback(e);
            isCompleted = true;
        });
    });
    if (self._opt.method === Constant.POST || self._opt.method === Constant.PUT) {
        req.write(querystring);
    }
    req.on('error', function (e) {
        if (isCompleted) return;
        clearTimeout(timer);
        e.message = 'Request failed. rid = ' + rid + ',interfaceId = ' + self._opt.id + ', url = '
        + self._opt.currUrl + '. Caused By: ' + e.message;
        errCallback(e);
        isCompleted = true;
    });

    req.end();
};

// @override interceptRequestReal function of ProxyBase which is defined in proxybase.js
HttpProxy.prototype.interceptRequestReal = function (req, res) {
    var qstr = req.url.replace(/^[^\?]*\??/, '');
    var params = '';
    var body;
    params = qs.parse(qstr);
    var self = this;
    var options = {
        hostname: self._opt.hostname,
        port: self._opt.port,
        path: self._parsePath(self._opt.path, params) + req.url.replace(/^[^\?]*\??/, ''),
        method: self._opt.method,
        headers: req.headers,
        agent: keepAliveAgent,
        keepAlive: true
    };

    options.headers.host = self._opt.hostname;
    // console.log( 'url2:\n', options.hostname + options.path );
    // delete options.headers.referer;
    // delete options.headers['x-requested-with'];
    // delete options.headers['connection'];
    // delete options.headers['accept'];
    delete options.headers['accept-encoding'];

    var req2 = http.request(options, function (res2) {
        var source = [], size = 0;

        res2.on('data', function (chunk) {
            source.push(chunk);
            size += chunk.length;
        });

        res2.on('end', function () {
            var buffer = Buffer.concat(source, size);
            var result;
            try {
                result = self._opt.encoding === Constant.ENCODING_RAW
                    ? buffer
                    : iconv.fromEncoding(buffer, self._opt.encoding);
            } catch (e) {
                res.statusCode = 500;
                res.end(e + '');
                return;
            }
            res.setHeader('Set-Cookie', res2.headers['set-cookie']);
            res.statusCode = res2.statusCode;
            res.setHeader('Content-Type', res2.headers['Content-Type']);
            res.end(result);
        });
        res2.on('error', function (err) {
            res.statusCode = 500;
            res.end(e + '');
        });
    });

    req2.on('error', function (e) {
        res.statusCode = 500;
        res.end(e + '');
    });
    req.on('data', function (chunck) {
        console.log(chunck.toString())
        req2.write(chunck);
    });
    req.on('end', function () {
        req2.end();
    });
};

HttpProxy.prototype._queryStringify = function (params) {
    if (!params || typeof params === 'string') {
        return params || '';
    } else if (params instanceof Array) {
        return params.join('&');
    }
    var qs = [], val;
    for (var i in params) {
        try {
            // to avoid circle reference of json
            val = typeof params[i] === 'object'
                ? JSON.stringify(params[i])
                : params[i];
        } catch (e) {
            val = params[i];
        }
        try {
            //  to avoid URIError: malformed URI sequence
            qs.push(i + '=' + encodeURIComponent(val));
        } catch (e) {
            console.log(new Date().toLocaleString()
            + '- URIError: malformed string = ' + val
            + ', from = ' + i
            + ', interfaceId = ' + this._opt.id
            + ', url = ' + this._opt.currUrl);

            qs.push(i + '=' + val);
        }
    }
    return qs.join('&');
};

/**
 * {Function}_parsePath
 * eg:
 * path = 'projects/:id/files/:name-:number?p1=vala&p2=:val'
 * params = {id:12345, name:'zyz', number:678, val:valb}
 * result => 'projects/12345/files/zyz-678?p1=vala&p2=valb'
 */
HttpProxy.prototype._parsePath = function (path, params) {
    if (!path || typeof path !== 'string') return path;
    if (typeof params !== 'object') return path;
    return path.replace(/\:\w+/g, function (s) {
        var key =  s.substring(1);
        var value =  params[key];
        delete params[key];
        return value;
    });
};

/**
 * {Function} init, will be called when the interface.json is loaded.
 * @param {Object} config The config of interface.json.
 */
HttpProxy.init = function (config) {
    httpConfig = config.http || {};
    keepAliveAgent = new Agent({
        maxSockets: httpConfig.maxSockets || 1000,
        keepAlive: true,
        keepAliveMsecs: httpConfig.keepAliveMsecs || 3000
    });
};

/**
 * {Function} verify will be called when the InterfaceManager is adding http interface profile and need
 * corresponding interface proxy implementer to verify whether the profile is valid.
 * @param {Object} prof The profile to be verified.
 * @param {Object} InterfaceManager This Object provides some useful methods to get some global
 * configuration such as getStatus so that some value of profile can be set as default.
 * @return {Object} prof The verified profile object.
 */
HttpProxy.verify = function (prof, InterfaceManager) {
    if (!(prof.status in prof.urls
        || prof.status === Constant.STATUS_MOCK
        || prof.status === Constant.STATUS_MOCK_ERR )) {
        prof.status = InterfaceManager.getStatus();
    }

    function isUrlsValid(urls) {
        if (!urls) return false;
        for (var i in urls) {
            return true;
        }
        return false;
    }

    if (!isUrlsValid(prof.urls)
        && !fs.existsSync(prof.ruleFile)) {
        throw new Error('Profile is deprecated:\n',
            prof, '\nNo urls is configured and No ruleFile is available');
    }

    prof.engine = prof.engine || InterfaceManager.getEngine();
    prof.method = {POST: 'POST', GET: 'GET', PUT: 'PUT'}
        [(prof.method || 'GET').toUpperCase()];
    prof.dataType = {json: 'json', text: 'text', jsonp: 'jsonp'}
        [(prof.dataType || 'json').toLowerCase()];
    prof.isRuleStatic = !!prof.isRuleStatic || false;
    prof.isCookieNeeded = !!prof.isCookieNeeded || false;
    prof.timeout = prof.timeout || httpConfig.keepAliveMsecs || 5000;
    prof.version = prof.version || '';
    prof.bypass = !!prof.bypassProxyOnClient;
    prof.url = ( prof.status !== Constant.STATUS_MOCK
    && prof.status !== Constant.STATUS_MOCK_ERR )
        ? prof.urls[prof.status] : '';

    if (prof.status === Constant.STATUS_MOCK || prof.status === Constant.STATUS_MOCK_ERR) {
        prof.bypass = false;
        // prof.dataType = prof.dataType === Constant.JSONP ? Constant.JSON : Constant.TEXT;
    }

    return prof;
};

module.exports = HttpProxy;